internal/statemachine/helper.go

```diff
@@ -1012,42 +1022,121 @@ func (stateMachine *StateMachine) updateGrub(rootfsVolName string, rootfsPartNum
+		},
 	}
 
-	for _, mp := range mountPoints {
-		mountCmds, umountCmds, err := mp.getMountCmd()
-		if err != nil {
-			return fmt.Errorf("Error preparing mountpoint \"%s\": \"%s\"",
-				mp.relpath,
-				err.Error(),
-			)
-		}
-		updateGrubCmds = append(updateGrubCmds, mountCmds...)
-		teardownCmds = append(umountCmds, teardownCmds...)
+	mountCmds, umountCmds, err := generateMountPointCmds(mountPoints, stateMachine.tempDirs.scratch)
+	if err != nil {
+		return err
 	}
 
+	setupGrubCmds = append(setupGrubCmds, mountCmds...)
+	teardownCmds = append(umountCmds, teardownCmds...)
+
 	teardownCmds = append([]*exec.Cmd{
 		execCommand("udevadm", "settle"),
 	}, teardownCmds...)
 
+	// udev needed to have grub-install properly work
+	grubPackages = append(grubPackages, "udev")
+
+	setupGrubCmds = append(setupGrubCmds,
+		aptInstallCmd(mountDir, grubPackages, false),
+		execCommand("chroot",
+			mountDir,
+			"grub-install",
+			loopUsed,
+			"--boot-directory=/boot",
+			"--efi-directory=/boot/efi",
+			fmt.Sprintf("--target=%s", target),
+			"--uefi-secure-boot",
+			"--no-nvram",
+		),
+	)
+
+	if architecture == arch.AMD64 {
+		setupGrubCmds = append(setupGrubCmds,
+			execCommand("chroot",
+				mountDir,
+				"grub-install",
+				loopUsed,
+				"--target=i386-pc",
+			),
+		)
+	}
+
 	divert, undivert := divertOSProber(mountDir)
 
-	updateGrubCmds = append(updateGrubCmds, divert)
+	setupGrubCmds = append(setupGrubCmds, divert)
 	teardownCmds = append([]*exec.Cmd{undivert}, teardownCmds...)
 
-	// actually run update-grub
-	updateGrubCmds = append(updateGrubCmds,
+	setupGrubCmds = append(setupGrubCmds,
 		execCommand("chroot",
 			mountDir,
 			"update-grub",
 		),
 	)
 
 	// now run all the commands
-	err = helper.RunCmds(updateGrubCmds, stateMachine.commonFlags.Debug)
+	return helper.RunCmds(setupGrubCmds, stateMachine.commonFlags.Debug)
+}
+
+func (stateMachine *StateMachine) confFromArch(architecture string) (string, []string) {
+	switch architecture {
+	case arch.AMD64:
+		return "x86_64-efi", []string{"grub-pc", "shim-signed"}
+	case arch.ARM64:
+		return "arm64-efi", []string{"grub-efi-arm64", "grub-efi-arm64-bin"}
+	case arch.ARMHF:
+		return "arm-efi", []string{"grub-efi-arm", "grub-efi-arm-bin"}
+	default:
+		return "", nil
+	}
+}
+
+// prepareGrubMountDir prepares a directory to run the grub installation process
+func prepareGrubMountDir(mountDir string, rootfsPartNum int, bootPartNum int, loopUsed string, debug bool) ([]*exec.Cmd, error) {
+	bootDir := filepath.Join(mountDir, "boot", "efi")
+	// Slice used to store all the commands that need to be run
+	// first to properly prepare the chroot
+	// setupGrubCmds should be filled as a FIFO list
+	var prepareCmds []*exec.Cmd
+	teardownCmds := []*exec.Cmd{
+		execCommand("umount", bootDir),
+		execCommand("umount", mountDir),
+	}
+	prepareCmds = append(prepareCmds,
+		// Try to make sure udev is not racing with losetup and briefly
+		// vanishing device files. See LP: #2045586
```

Comment:
I followed the fix and discussions about that when it was fixed in livecd-rootfs. I would like to fix it properly everywhere (this is not the only place we are calling `udevadm`) but this PR was not the place to do it, especially because in need to experiment with `flock` and check if this is available on older series. See FR-6372

Reply:
flock has been part of util-linux (which is Essential) for about a million years (I think it might have been new in 5.04?) but ok